### PR TITLE
Fixed TD(lambda) update, added T-Maze

### PR DESCRIPTION
**Pull request by taodav**
Quite a few changes made here!

First is to fix a bug in the TD(lambda) implementation, where the bootstrapped targets were using the same action as the current action. Changed that to the next action.

Second is to add Bakker's T-Maze as an example!

Third, I've turned all the calls to `jax.numpy` back to `numpy` calls (we see a big speed increase from this). If experiments every get unwieldy I can always look into `jit`ting things to speed them up even more. 

Fourthly (idk if that's a word), I've added a small test for the TD(lambda) implementation with the Simple Chain environment.
The Simple Chain environment is a 10-state MDP (with all states connected in a line) with a single action (go right), and a single reward from the last non-terminal state to the terminal state. The agent always initializes in the left-most state, and simply goes towards the terminal state in 9 steps. The ground-truth values that all methods should learn is simple to calculate. I might append onto this PR to also test the analytical solving method.

Lastly, I've added a `setup.py` file and added `-e .` to `requirements.txt`. This installs the `grl` package locally, allowing for edits. This is to also get rid of some of the relative imports in favor of calling something like `grl.environment` etc.

Make sure you run `pip install -e .` if this commit goes through!

### CODE REVIEW

**Comment by camall3n** [comment]
Does changing the imports from relative to absolute also change the way we need to run stuff? Does everything still work as described in the README?

**Comment by dsctt** [comment]
Noting that if it's significantly faster, then changing back to normal numpy sounds good but it just means that if anyone is ever doing gradient stuff they'll have to switch it back to jax.numpy at that time (or figure out something clever).

**Comment by dsctt** [comment]
I'm going to merge the other PR that's been up for a while first so we can stay in sync. It's pretty big and they touch the same files. (Could've been merged a while ago but there wasn't anything else going on in the repo.)

File: tests/test_tmaze.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+if __name__ == "__main__":
+    corridor_length = 0
+    T, R, gamma, p0, phi = tmaze(corridor_length)
+    mdp = MDP(T, R, p0, gamma=gamma)
+    pomdp = AbstractMDP(mdp, phi)
+
+    n_samples = 10000
+
+    # first we test our start states
```
Comment (dsctt) [inline tests/test_tmaze.py:15]: Can you make the tests in both files like the ones in e.g. `test_policy_eval.py`? That way they run and show up nicely in pytest (and can be separated out too if it seems helpful)

File: grl/environment/tmaze_lib.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    n_states += 2  # T junctions
+    n_states += 1  # terminal state
+
+    T_up = np.eye(n_states, n_states)
+    T_down = T_up.copy()
+    T_up[[-1, -2, -3], [-1, -2, -3]] = 0
+    T_down[[-1, -2, -3], [-1, -2, -3]] = 0
+
+    # If we go up or down at the junctions, we terminate
+    T_up[[-1 - 1, -2 - 1], [-1, -1]] = 1
```
Comment (dsctt) [inline grl/environment/tmaze_lib.py:19]: Should we keep the subtraction? I think you are showing it to show how we work backwards to the T junction states, but then wouldn't it be `[-1 - 1, -1 - 2]` ? I'm not quite sure how it is making it more clear tbh.

File: grl/environment/tmaze_lib.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    # At the leftmost and rightmost states we transition to ourselves
+    T_left[[0, 1], [0, 1]] = 1
+    T_right[[-1 - 1, -2 - 1], [-1 -1, -2 - 1]] = 1
+
+    # transition to -2 (left) or +2 (right) index
+    all_nonterminal_idxes = np.arange(n_states - 1)
+    T_left[all_nonterminal_idxes[2:], all_nonterminal_idxes[:-2]] = 1
+    T_right[all_nonterminal_idxes[:-2], all_nonterminal_idxes[2:]] = 1
+
+    T = np.array([T_up, T_right, T_down, T_left])
```
Comment (dsctt) [inline grl/environment/tmaze_lib.py:34]: I'm partial to `[T_up, T_down, T_right, T_left]` and same for R because when there are 2 actions it would be `[T_up, T_down]` (or `[T_right, T_left]`) and it'll be more natural to read it this way with 4 actions. Guess it's not really a big deal tho.

File: grl/environment/tmaze_lib.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+    R = np.array([R_up, R_right, R_down, R_left])
+
+    # Initialize uniformly at random between north start and south start
+    p0 = np.zeros(n_states)
+    p0[:2] = 0.5
+
+    # Observation function. We have 4 possible obs - start_up, start_down, corridor and junction.
+    # terminal obs doesn't matter.
+    phi = np.zeros((n_states, 4))
```
Comment (dsctt) [inline grl/environment/tmaze_lib.py:61]: In `examples_lib` we do include terminal obs by convention. If it doesn't matter it doesn't matter, just noting in case it comes up somehow.

File: grl/environment/tmaze_lib.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+    phi[0, 0] = 1
+    phi[1, 1] = 1
+
+    phi[2:2 + n, 2] = 1
```
Comment (dsctt) [inline grl/environment/tmaze_lib.py:66]: Why doesn't this work instead of those 3 lines: `phi[2:-3, 2]=1`?

**Review by dsctt — CHANGES_REQUESTED** [review]
Awesome. Few questions and comments.

**Comment by dsctt** [comment]
> Noting that if it's significantly faster, then changing back to normal numpy sounds good but it just means that if anyone is ever doing gradient stuff they'll have to switch it back to jax.numpy at that time (or figure out something clever).

Wait also the jax `.at` and `.set` calls don't translate back to old numpy. So policy_eval etc. would break. Would have to change those too if we want to revert back to numpy.

**Comment by camall3n** [comment]
> Noting that if it's significantly faster, then changing back to normal numpy sounds good but it just means that if anyone is ever doing gradient stuff they'll have to switch it back to jax.numpy at that time (or figure out something clever).

Maybe we should have a switch that enables jax only when necessary

**Comment by camall3n** [comment]
> > Noting that if it's significantly faster, then changing back to normal numpy sounds good but it just means that if anyone is ever doing gradient stuff they'll have to switch it back to jax.numpy at that time (or figure out something clever).
> 
> Wait also the jax `.at` and `.set` calls don't translate back to old numpy. So policy_eval etc. would break. Would have to change those too if we want to revert back to numpy.

Can we wrap that stuff in functions that know whether we're using jax?

File: grl/environment/tmaze_lib.py (hunk 5 above)
Comment (camall3n) [reply grl/environment/tmaze_lib.py:66]: Lines 63 and 64 need to have their own distinct observations. The remaining states all get the same observation, and there are `n+2` of them. But you could still write line 66 more simply as `phi[2:, 2] = 1`.

File: tests/test_tmaze.py (hunk 1 above)
Comment (camall3n) [reply tests/test_tmaze.py:15]: If you have tests that need to be configured in the same way, you can use [pytest fixtures](https://docs.pytest.org/en/6.2.x/fixture.html#what-fixtures-are) to accomplish that.

Here's an example of how that works:
https://github.com/camall3n/visgrid/blob/master/test/test_gridworld.py

**Comment by taodav** [comment]
@dsctt I've discussed with Cam, and I think what makes more sense right now is to push a separate PR for a fix and test for TD(lambda), and have a lot of this as a separate thing along with some of the JAX `jit`ing I want to add. I'll make a separate PR with all the changes for getting JAX working faster and the T-Maze stuff.
